vm bridge for erc20/cw20

BananaLF opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰    
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

- Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/okex/okexchain/blob/develop/CONTRIBUTING.md#pr-targeting))

- [ ] Linked to github-issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Wrote tests
- [ ] Updated relevant documentation (`docs/`)
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Reviewed `Files changed` in the github PR explorer

______

For Admin Use:
- Added appropriate labels to PR (ex. wip, ready-for-review, docs)
- Reviewers Assigned
- Squashed all commits, uses message "Merge PR #XYZ: [title]" ([coding standards](https://github.com/okex/coding/blob/master/README.md#merging-a-pr))

BananaLF committed:
init bridge

BananaLF committed:
commplete bridge

BananaLF committed:
Merge branch 'dev' into sunstrider/bridge

BananaLF committed:
add abi file

BananaLF committed:
Merge remote-tracking branch 'origin/sunstrider/bridge' into sunstrider/bridge

BananaLF committed:
add rust code

BananaLF committed:
add unit test

BananaLF committed:
add test erc20 contract

BananaLF committed:
disable genesis for wasm

BananaLF committed:
Merge branch 'dev' into sunstrider/bridge

# Conflicts:
#	app/app.go

BananaLF committed:
disable genesis for wasm

BananaLF committed:
change err msg for update wasm contract

BananaLF committed:
op wasm return

BananaLF committed:
Merge branch 'dev' into sunstrider/bridge

xzav3r committed:
Merge branch 'dev' into sunstrider/bridge

BananaLF committed:
Merge branch 'dev' into sunstrider/bridge

BananaLF committed:
change Venus3 to Earth

BananaLF committed:
change Venus3 to Earth

BananaLF committed:
change Venus3 to Earth

BananaLF committed:
make unit test pass

BananaLF committed:
solve a bug for proposal must be earth height

BananaLF committed:
Merge branch 'dev' into sunstrider/bridge

xzav3r commented on x/vmbridge/types/protocol.go:
```diff
@@ -0,0 +1,118 @@
+)
+
+func init() {
+	//stringType, _ := abi.NewType("string", "", nil)
```
code annotations shoud be deleted if unused

xzav3r commented on x/vmbridge/types/tx.go:
```diff
@@ -0,0 +1,57 @@
+	if err != nil {
+		return err
+	}
+	if sdk.IsWasmAddress(recipient) {
```
Why recipient should be wasm address? or it should be a evm address?

BananaLF replied:
MsgSendToEvm must be call evmContract to transfer erc20 to user. In Evm contract may only support evm address.

xzav3r commented on x/wasm/types/tx.go:
```diff
@@ -220,7 +220,7 @@ func (msg MsgUpdateAdmin) ValidateBasic() error {
 	}
 	if strings.EqualFold(msg.Sender, msg.NewAdmin) {
-		return sdkerrors.Wrap(ErrInvalidMsg, "new admin is the same as the old")
+		return sdkerrors.Wrap(ErrInvalidMsg, "new admin is the same as the old admin (you input)")
```
the user input the new admin, is it neccessary to modify the error message?

BananaLF replied:
If contract is empty，then return “new admin is the same as the old”. it maybe cause user blinded. so i change errmsg for user make it understand errmsg.

xzav3r commented on x/wasm/module_adapter.go:
```diff
@@ -33,18 +33,11 @@ func (b AppModuleBasic) RegisterCodec(amino *codec.Codec) {
-		return err
-	}
-	return ValidateGenesis(data)
+	return nil
```
validate should not be deleted because it can be used in other places

BananaLF replied:
I make wasm the same as ibc module. And Validate Genesis is only called by `CMD`. This code delete is for wasm is not supported in genensis file

xzav3r commented on x/vmbridge/keeper/evm.go:
```diff
@@ -0,0 +1,134 @@
+	}
+
+	logger := h.Keeper.Logger()
+	unpacked, err := types.SendToWasmEvent.Inputs.Unpack(data)
```
is there any condition that the length of `unpacked` is not 3 or the filed type of it if not `string, string, *big,Int`? Maybe it is better to check it and add some annotations

BananaLF replied:
It's nice suggestion，i will add more unit test to check it. But Input.Unpack is ETH abi interface of go language, it's risk not high

BananaLF replied:
I add some unit test. fake data ( bytes,string,big.int) it can be normal unpack, but after will check
 address. And fake data (string,string,int8(-1)) it also can be normal unpack, but is panic in sdk.NewIntFromBigInt  it's correct action. 
The most importante case in func (lh LogProcessEvmHook) PostTxProcessing it will search event.ID to handle different logic. event.ID = hash(event.names, events.inputs.type)。So we can regard Unpack.data is must be correct type。

BananaLF committed:
Merge branch 'dev' into sunstrider/bridge

BananaLF committed:
fix code review

ylsGit committed:
Merge branch 'dev' into sunstrider/bridge

BananaLF committed:
make unit test pass

BananaLF committed:
Merge remote-tracking branch 'origin/sunstrider/bridge' into sunstrider/bridge

ylsGit committed:
Merge branch 'dev' into sunstrider/bridge

BananaLF committed:
Merge branch 'dev' into sunstrider/bridge

BananaLF committed:
add vmbridge control feild into wasm.params

BananaLF committed:
change err code

BananaLF committed:
Merge remote-tracking branch 'origin/sunstrider/bridge' into sunstrider/bridge

BananaLF committed:
make unit test pass

BananaLF committed:
solve bug for wasm.param cache

BananaLF committed:
solve bug for wasm cache that we can not cost gas within area which is used cache

ylsGit commented on x/vmbridge/keeper/wasm.go:
```diff
@@ -0,0 +1,94 @@
+		return types.ErrIsNotWasmAddr
+	}
+	params := k.wasmKeeper.GetParams(ctx)
+	if !params.VmbridgeEnable {
```
Recommended for implementation in SendToWasmEventHandler

BananaLF committed:
change vmbridge check in evm

BananaLF committed:
Merge branch 'dev' into sunstrider/bridge

BananaLF committed:
disable params cache

BananaLF committed:
make unit test pass

ylsGit committed:
Merge branch 'dev' into sunstrider/bridge

BananaLF committed:
check address:
1. sendtowasm, to must okc address
2. sendtoevm, to.addr/contract.addr must be eth address

BananaLF committed:
change error msg

BananaLF committed:
change error msg

BananaLF committed:
change error msg

BananaLF committed:
Merge branch 'dev' into sunstrider/bridge

ylsGit committed:
Merge branch 'dev' into sunstrider/bridge

BananaLF committed:
Merge branch 'dev' into sunstrider/bridge

xiangjianmeng committed:
Merge branch 'dev' into sunstrider/bridge
